task/universal/environment.go

```diff
@@ -0,0 +1,99 @@
+package universal
+
+import (
+	"encoding/base64"
+	"fmt"
+	"os"
+	"strings"
+	"time"
+)
+
+type Environment struct {
+	Image     string
+	Script    string
+	Variables map[string]*string
+	Timeout   time.Duration
+	Directory string
+}
+
+func (e *Environment) GenerateMachineScript() string {
+	var variables string
+	for name, value := range e.Variables {
+		if value == nil {
+			data := os.Getenv(name)
+			value = &data
+		}
+		escaped := strings.ReplaceAll(*value, `"`, `\"`) // FIXME: \" edge cases.
+		variables += fmt.Sprintf("%s=\"%s\"\n", name, escaped)
+	}
+
+	script := fmt.Sprintf(
+		`#!/bin/bash
+sudo mkdir --parents /tmp/tpi-task
+chmod u=rwx,g=rwx,o=rwx /tmp/tpi-task
+
+base64 --decode << END | sudo tee /usr/bin/tpi-task > /dev/null
+%s
+END
+chmod u=rwx,g=rx,a=rx /usr/bin/tpi-task
+
+base64 --decode << END | sudo tee /tmp/tpi-environment > /dev/null
+%s
+END
+chmod u=rw,g=,o= /tmp/tpi-environment
+
+while IFS= read -rd $'\0' variable; do
+  export "$(perl -0777p -e 's/\\"/"/g;' -e 's/(.+?)="(.+)"/$1=$2/sg' <<< "$variable")"
+done < <(perl -0777pe 's/\n*(.+?=".*?((?<!\\)"|\\\\"))\n*/$1\x00/sg' /tmp/tpi-environment)
+
+sudo tee /etc/systemd/system/tpi-task.service > /dev/null <<END
+[Unit]
+  After=default.target
+[Service]
+  Type=simple
+  ExecStart=/usr/bin/tpi-task
+  ExecStop=/bin/bash -c 'systemctl is-system-running | grep stopping || tpi --stop'
+  EnvironmentFile=/tmp/tpi-environment
+  WorkingDirectory=/tmp/tpi-task
+  TimeoutStartSec=%d
+[Install]
+  WantedBy=default.target
+END
+
+curl --location --remote-name https://github.com/iterative/terraform-provider-iterative/releases/latest/download/terraform-provider-iterative_linux_amd64
+sudo mv terraform-provider-tpi* /usr/bin/tpi
+sudo chmod u=rwx,g=rx,o=rx /usr/bin/tpi
+sudo chown root:root /usr/bin/tpi
+
+curl --remote-name https://downloads.rclone.org/rclone-current-linux-amd64.zip
+python3 -m zipfile -e rclone-current-linux-amd64.zip .
+sudo cp rclone-*-linux-amd64/rclone /usr/bin
+sudo chmod u=rwx,g=rx,o=rx /usr/bin/rclone
+sudo chown root:root /usr/bin/rclone
+rm --recursive rclone-*-linux-amd64*
+
+rclone copy "$RCLONE_REMOTE/data" /tmp/tpi-task
+
+sudo systemctl daemon-reload
+sudo systemctl enable tpi-task.service --now
+
+TPI_IDENTITY="$(uuidgen)"
+
+while sleep 5; do
```

Comment:
maybe this 5 secs update might be configurable

Reply:
Not now, at least. This code is subject to change, and public API is forever.

Reply:
Not for the user, I mean for us... Its 5 secs a good election? also if a good reminder

Reply:
Maybe not a good election: if `rclone` is dumb enough (untested), S3 PUT operations cost 0.005 USD each and that means 86.4 USD per day in log writes. 🙈 

What if we optimize it so it writes only when there is a new log line? It will cost more for log bursts but less for long–running tasks.

```suggestion
journalctl --unit tpi-task | while read line; do
```

Ideally it should be a mix of both: refresh every five seconds and only push of the contents are different. Trivial to implement, but still an optimization.

Reply:
Thats what I say, why dont we expose does parameters to be finetuned?

Reply:
Because it's not an hyperparameter search issue, but rather a design one. I'll follow up with The Only Valid Approach™ as a separate optimization.